### PR TITLE
Inactive Tabs: FXIOS-1738

### PR DESCRIPTION
**Pull request by nbhasin2**
Implemented inactive tabs for iPad and iPhone

- When tabs are inactive for more than 4 days we send them to inactive list 
- When tabs are not active for 30days we remove them
- Added support for Inactive tabs to not work in private mode 
- Added recently closed section in tab tray for inactive tabs
- Added nimbus support for experimentation 
- Added debug menu to enable / disable inactive tabs
- Fixed rotation issue on iPad where tab sizes were incorrect 
- Did cleanup to remove some un-used icons from the app
- Recently closed now holds 10 tabs instead of 5 that it used to hold before
- Added first created tab time so that we can better calculate when to put tabs into inactive mode when there is no session data or last used time
- Added telemetry events for
a) expand / collapse inactive tab tray
b) open an inactive tab
c) open recently closed list from inactive tab tray
d) open a recently closed tab

Ref: FXIOS-1738, 1942, 1943, 1877, 1878

Simulator video showing what happens to inactive tab when I change the date in the background using system settings.

https://user-images.githubusercontent.com/8919439/124804345-307b3980-df28-11eb-824f-532390222e46.mov

### CODE REVIEW

**Comment by nbhasin2** [comment]
This is an MR2 work so it should not be merged before v35 

Remaining: 

a) Nimbus experiment support on Backend 
b) Rigorous testing on iPad as it has floating tabs + grid layout which can likely have issues syncing when experiment is turned on while user is using the app

Still debating if I should just enable full feature on next launch 🤔

**Comment by nbhasin2** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?

    How many users: 
    a) expand / collapse inactive tab tray
    b) open an inactive tab 
    c) open recently closed list from inactive tab tray
    d) open a recently closed tab 

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:
    - These events will help us understand the usage of inactive tabs tray and recently closed tabs section. This will eventually help us make product decision on how to improve inactive tabs tray feature.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

   None

4) Can current instrumentation answer these questions?

   No

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

Category 2

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>

  <tr>
    <td>Counts the number of times user toggle (expand or collapse)
      inactive tab section.</td>
    <td>Category 2</td>
    <td>https://mozilla-hub.atlassian.net/browse/FXIOS-1738</td>
  </tr>

  <tr>
    <td>Counts the number of times user opens an inactive tab.</td>
    <td>Category 2</td>
    <td>https://mozilla-hub.atlassian.net/browse/FXIOS-1738</td>
  </tr>

  <tr>
    <td>Counts the number of times user opens recently closed
      section from inactive tabs tray.</td>
    <td>Category 2</td>
    <td>https://mozilla-hub.atlassian.net/browse/FXIOS-1942</td>
  </tr>

  <tr>
    <td>Counts the number of times the Go To Settings button on
      default browser card is clicked.</td>
    <td>Category 2</td>
    <td>https://mozilla-hub.atlassian.net/browse/FXIOS-1877</td>
  </tr>

</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
Available in metrics.yaml and https://dictionary.telemetry.mozilla.org/apps/firefox_ios

7) How long will this data be collected?  Choose one of the following:

    Until 2022-01-08 (approximately 6 months) so we can gather enough data

8) What populations will you measure?
All

9) If this data collection is default on, what is the opt-out mechanism for users?
Default on, users can opt out in settings

10) Please provide a general description of how you will analyze this data.
Using Glean / Amplitude

11) Where do you intend to share the results of your analysis?
Only on Glean, Amplitude and with mobile teams.

12) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:
None

**Comment by dnarcese** [comment]
I think the last description in the measurement chart needs to be updated, otherwise r+

# Data Review Form

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? 

Yes, telemetry is documented in `docs/metrics.md`

2) Is there a control mechanism that allows the user to turn the data collection on and off? 

Yes, Firefox iOS provides data controls in the app settings

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

This should be re-reviewed w/ Ashley Thomas every 6mo

4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Type 2

5) Is the data collection request for default-on or default-off?

Default on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

No

7) Is the data collection covered by the existing Firefox privacy notice? 

Yes

8) Does the data collection use a third-party collection tool? 

No

**Comment by nbhasin2** [comment]
@dnarcese oops, I updated last description for opening recently closed tab

File: Client/Frontend/Browser/TabDisplayManager.swift [hunk 1]
```diff
@@ -145,7 +202,7 @@ class TabDisplayManager: NSObject {
     func indexOfCellDrawnAsPreviouslySelectedTab(currentlySelected: Tab) -> IndexPath? {
         for i in 0..<collectionView.numberOfItems(inSection: 0) {
             if let cell = collectionView.cellForItem(at: IndexPath(row: i, section: 0)) as? TopTabCell, cell.selectedTab {
-                if let tab = dataStore.at(i), tab != currentlySelected {
+                if let tab = dataStore.at(i) , tab != currentlySelected {
```
Comment (electricRGB) [inline Client/Frontend/Browser/TabDisplayManager.swift:205]:
```suggestion
                if let tab = dataStore.at(i), tab != currentlySelected {
```

File: Client/Frontend/Browser/Tabs/InactiveTabCell.swift [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+    }()
+
+    convenience init(viewModel: InactiveTabViewModel) {
+        self.init()
+        inactiveTabsViewModel = viewModel
+    }
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        tableView.dataSource = self
```
Comment (electricRGB) [inline Client/Frontend/Browser/Tabs/InactiveTabCell.swift:54]: How come we're doing this in `layoutSubviews`?

**Review by electricRGB — APPROVED** [review]
I've tested as I could and everything seems mostly ok for me. I think there'll probably be some issues that come up that we're just not seeing, but I can't find anything toooo wonky.

**Pull request merged**